### PR TITLE
fix buffer gc wrong index on metal

### PR DESCRIPTION
**Pull request by hana-alice**
Re: #https://github.com/cocos/3d-tasks/issues/13682

### Changelog

*

-------

### Continuous Integration

This pull request:

* [ ] needs automatic test cases check.
  > Manual trigger with `@cocos-robot run test cases` afterward.
* [ ] does not change any runtime related code or build configuration
  > If any reviewer thinks the CI checks are needed, please uncheck this option, then close and reopen the issue.

-------

### Compatibility Check

This pull request:

* [ ] changes public API, and have ensured backward compatibility with [deprecated features](https://github.com/cocos/cocos-engine/blob/v3.5.0/docs/contribution/deprecated-features.md).
* [ ] affects platform compatibility, e.g. system version, browser version, platform sdk version, platform toolchain, language version, hardware compatibility etc.
* [ ] affects file structure of the build package or build configuration which requires user project upgrade.
* [ ] **introduces breaking changes**, please list all changes, affected features and the scope of violation.

### CODE REVIEW

**Comment by hana-alice** [comment]
@cocos-robot run test cases

File: native/cocos/renderer/gfx-metal/MTLDevice.mm [hunk 1]
```diff
@@ -232,21 +234,21 @@ of this software and associated engine source code (the "Software"), a limited,
...
+    if (index >= 0 && index < MAX_FRAMES_IN_FLIGHT) {
+        CCMTLGPUStagingBufferPool *bufferPool = _gpuStagingBufferPools[index];
         if (bufferPool) {
             bufferPool->reset();
-            CCMTLGPUGarbageCollectionPool::getInstance()->clear(_currentBufferPoolId);
+            CCMTLGPUGarbageCollectionPool::getInstance()->clear(index);
         }
     }
-    _inFlightSemaphore->signal();
+    _inFlightSemaphores[index]->signal();
```
Comment (bofeng-song) [inline native/cocos/renderer/gfx-metal/MTLDevice.mm]: whether index may exceed the size of the array

Comment (hana-alice) [reply native/cocos/renderer/gfx-metal/MTLDevice.mm]: index = ++index % 3 => index ranges {0, 2};
inFlightSemaphores.size() == 3. 
so it won't access out side of the array.

File: native/cocos/renderer/gfx-metal/MTLGPUObjects.h [hunk 2]
```diff
@@ -199,45 +192,23 @@ class CCMTLGPUStagingBufferPool final {
...
         }
         gpuBuffer->mtlBuffer = buffer->mtlBuffer;
         gpuBuffer->startOffset = offset;
         gpuBuffer->mappedData = buffer->mappedData + offset;
         buffer->curOffset = offset + gpuBuffer->size;
+        
     }
 
     void updateInflightBuffer() {
-        if (_tripleEnabled) {
```
Comment (minggo) [inline native/cocos/renderer/gfx-metal/MTLGPUObjects.h:228]: It seems this function is empty, can remove it?

File: native/cocos/renderer/gfx-metal/MTLGPUObjects.h [hunk 3]
```diff
@@ -199,45 +192,23 @@ class CCMTLGPUStagingBufferPool final {
             }
         }
         if (!buffer) {
-            uint32_t mbNeeds = mu::roundUp(gpuBuffer->size, MegaBytesToBytes);
-            mbNeeds = cc::utils::nextPOT(mbNeeds);
-            CC_ASSERT(mbNeeds > gpuBuffer->size / static_cast<float>(MegaBytesToBytes));
+            uint32_t needs = mu::alignUp(gpuBuffer->size, MegaBytesToBytes);
```
Comment (minggo) [inline native/cocos/renderer/gfx-metal/MTLGPUObjects.h:195]: Why change to logic?

File: native/cocos/renderer/gfx-metal/MTLDevice.mm [hunk 4]
```diff
@@ -75,7 +75,9 @@ of this software and associated engine source code (the "Software"), a limited,
 
 bool CCMTLDevice::doInit(const DeviceInfo &info) {
     _gpuDeviceObj = ccnew CCMTLGPUDeviceObject;
-    _inFlightSemaphore = ccnew CCMTLSemaphore(MAX_FRAMES_IN_FLIGHT);
+    for(size_t i = 0; i < MAX_FRAMES_IN_FLIGHT; ++i) {
+        _inFlightSemaphores[i] = ccnew CCMTLSemaphore(1);
+    }
```
Comment (minggo) [inline native/cocos/renderer/gfx-metal/MTLDevice.mm]: I think just use one semaphore is enough.

**Comment by hana-alice** [comment]
1. Remove _tripleBufferEnabled in the stagingBuffer, because there is already a staginBuffer array of size 3 in the device. The logic of trippleBuffer is also unified in the device as much as possible.
2. Device::Present puts an onPresentComplete callback in the commandBuffer, but the onPresentComplete executes _currentPoolID as the index. Under the logic of the tirpple buffer, if the gpu execution is stuck in the 0th frame, the tripple buffer is filled to the 2nd frame, the 0th frame The onPresentComplete should be to clean up the stagingBuffer of the 0th frame, here because _currentPoolID cleans up the 2nd frame. The callback should capture the stack-based index.
3. ~~Changed inFlightSemaphore to 3 indexes and currentFrameIndex. This piece just looks a little clearer.~~
update: In the case of 2, if keep this situation, the buffer 0 and buffer 1 have never been reset and only continue to alloc, thus the memory will continue to grow.

---

1. 移除掉的stagingBuffer里边的_tripleBufferEnabled, 因为device里边已经有一个大小为3的staginBuffer数组了。trippleBuffer的逻辑也尽量统一在device里边。
2. Device::Present在commandBuffer里边放了一个onPresentComplete回调，但是onPresentComplete 里边执行的是_currentPoolID作为index. tirpple buffer的逻辑下，如果gpu执行卡在第0帧，tripple buffer填充到了第2帧，第0帧的onPresentComplete应该是去清理第0帧的stagingBuffer，在这里就因为_currentPoolID清理的是第2帧了。回调因该捕捉stack-based的index.
3. ~~统一把inFlightSemaphore也改成3个index和currentFrameIndex匹配起来，分析下来这块只是看起来清晰一点。~~

update: 2的情况下如果一直保持这种节奏，0和1buffer从未被重置只有不断申请，是会出现内存一直不断增长的情况的。

**Comment by star-e** [comment]
I see. 
There are several points I want to mention.
1. "number of gpu frames" might not equal "number of backbuffers". In our current implementation, they are equal to 3. But in some configuration, number of frames might fewer than number of backbuffers. (e.g num of frames == 2 instead of 3). It is required frames==2 to achieve lowest latency in some situation.

2. User might allocate too much staging buffer at once. In the future, we need to throttle upload size in one frame.

3. Synchronization might happen mutiple-times per frame (Per-RenderPass), once per frame (Per-Frame, current situation), once several frames (Per-Task). Different usages might need different implementation.

我了解了，有几点需要注意，这里提一下。
1. 逻辑上的gpu frame数量与backbuffer数量，有时会不一致，比如frame == 2，backbuffer == 3。当frame==2时，才有可能达到最低的输入->显示延迟，这是必要条件。
2. 用户可能一次性上传大量数据/发布了太多上传任务。我们之后需要做一些限流，控制staging buffer总大小，平滑帧率。
3. 同步的频率，有很多种可能，并不一定是每帧一次。比如RenderPass一帧有多次，异步的上传/计算任务可能跨多帧。之后不同的情况需要不同的处理。

**Comment by hana-alice** [comment]
Got it.

**Review by minggo — APPROVED** [review]

**Review by bofeng-song — APPROVED** [review]

**Review by star-e — APPROVED** [review]

**Pull request merged**